ReactNativeClient/lib/services/searchengine/queryBuilder.ts

```diff
@@ -0,0 +1,410 @@
+const { time } = require('lib/time-utils.js');
+
+interface Term {
+	name: string
+	value: string
+	negated: boolean
+}
+
+enum Relation {
+    OR = 'OR',
+    AND = 'AND',
+}
+
+enum Operation {
+	UNION = 'UNION',
+	INTERSECT = 'INTERSECT'
+}
+
+enum Requirement {
+	EXCLUSION = 'EXCLUSION',
+	INCLUSION = 'INCLUSION'
+}
+
+const notebookFilter = (terms: Term[], conditions: string[], params: string[], withs: string[]) => {
+	const notebooks = terms.filter(x => x.name === 'notebook' && !x.negated).map(x => x.value);
+	if (notebooks.length === 0) return;
+
+	const likes = new Array(notebooks.length).fill('folders.title LIKE ?').join(' OR ');
+	const withInNotebook = `
+	notebooks_in_scope(id)
+	AS (
+		SELECT folders.id
+		FROM folders
+		WHERE id
+		IN (
+			SELECT id
+			FROM folders
+			WHERE ${likes}
+		)
+		UNION ALL
+		SELECT folders.id
+		FROM folders
+		JOIN notebooks_in_scope
+		ON folders.parent_id=notebooks_in_scope.id
+	)`;
+	const where = `
+	AND ROWID IN (
+		SELECT notes_normalized.ROWID
+		FROM notebooks_in_scope
+		JOIN notes_normalized
+		ON notebooks_in_scope.id=notes_normalized.parent_id
+	)`;
+
+	withs.push(withInNotebook);
+	params.push(...notebooks);
+	conditions.push(where);
+};
+
+const getOperator = (requirement: Requirement, relation: Relation): Operation => {
+	if (relation === 'AND' && requirement === 'INCLUSION') { return Operation.INTERSECT; } else { return Operation.UNION; }
+};
+
+const filterByTableName = (
+	terms: Term[],
+	conditions: string[],
+	params: string[],
+	relation: Relation,
+	noteIDs: string,
+	requirement: Requirement,
+	withs: string[],
+	tableName: string
+) => {
+	const operator: Operation = getOperator(requirement, relation);
+
+	const values = terms.map(x => x.value);
+
+	let withCondition = null;
+	// with_${requirement}_${tableName} is added to the names to make them unique
+	if (relation === Relation.OR && requirement === Requirement.EXCLUSION) {
+		withs.push(`
+		all_notes_with_${requirement}_${tableName}
+		AS (
+			SELECT DISTINCT note_${tableName}.note_id AS id FROM note_${tableName}
+		)`);
+
+		const notesWithoutExcludedField = `
+		SELECT * FROM (
+			SELECT *
+			FROM all_notes_with_${requirement}_${tableName}
+			EXCEPT ${noteIDs}
+		)`;
+
+		// We need notes without atleast one excluded tableName (tag/resource)
+		withCondition = `
+		notes_with_${requirement}_${tableName}
+		AS (
+			${new Array(values.length)
+		.fill(notesWithoutExcludedField)
+		.join(' UNION ')}
+		)`;
+
+	} else {
+		// Notes with any/all values depending upon relation and requirement
+		withCondition = `
+		notes_with_${requirement}_${tableName}
+		AS (
+			SELECT note_${tableName}.note_id as id
+			FROM note_${tableName}
+			WHERE
+			${operator === 'INTERSECT' ? 1 : 0} ${operator}
+			${new Array(values.length).fill(noteIDs).join(` ${operator} `)}
```

Comment:
Please also move this kind of computation outside the template to improve readability. Also please don't use the `new Array` constructor. In fact, I've disallowed it in eslint.rc (no-array-constructor rule) but somehow it's not picking up the error.